### PR TITLE
Add method to get an array of types a module is directly included in

### PR DESCRIPTION
**Pull request by Blacksmoke16**
```cr
module Foo
end

module Baz
  module Tar
    include Baz
  end
end

abstract class Parent
end

class Bar < Parent
  include Foo
  include Baz
end

struct Str
  include Baz
end

struct Gen(T)
  include Baz
end

abstract struct AStr
  include Baz
end

abstract class ACla
  include Baz
end

class SubT(T)
  include Baz
end

class ChildT(T) < SubT(T)
end

{{Baz.all_includers.map &.name.stringify}} # => ["Baz::Tar", "Bar", "Str", "Gen(T)", "AStr", "ACla", "SubT(T)"]
```

Questions:
* ~How do I test this?~
* ~Are there any other types that can include a module i should try?~
* ~Any other ideas for a better name?~

I also noticed that types with generics include the generic type vars in their name.  I would have thought `Gen(T)` would have just been `Gen`...not sure if this is intended or not.

EDIT:  I added it to `macros.cr`, will include that in feedback fixes.

### CODE REVIEW

**Comment by kalinon** [comment]
Awesome. this will be quite useful if merged. I cant think of anything off the top of my head that also includes modules.

**Comment by sam0x17** [comment]
idea: you could make a macro that generates a bunch of arbitrary modules to test

**Comment by asterite** [comment]
Any use case for this?

**Comment by Blacksmoke16** [comment]
Mainly as a compliment to `#subclassess`.  ATM only way to get an array of types is using that.  This allows for using modules to achieve a similar effect if inheritance isn't the best solution.

Ref https://forum.crystal-lang.org/t/type-all-includers/254

File: src/compiler/crystal/macros/methods.cr [hunk 1]
```diff
@@ -1654,6 +1656,17 @@ module Crystal
       end
     end
 
+    def self.including_types(type)
+      if type.is_a? NonGenericModuleType
```
Comment (asterite) [inline src/compiler/crystal/macros/methods.cr]: I think this logic also applies to `GenericModuleType` (but you'll have to do it in a separate branch with duplicate logic).

File: src/compiler/crystal/macros/methods.cr [hunk 2]
```diff
@@ -1522,6 +1522,8 @@ module Crystal
         interpret_argless_method(method, args) { TypeNode.subclasses(type) }
       when "all_subclasses"
         interpret_argless_method(method, args) { TypeNode.all_subclasses(type) }
+      when "all_includers"
+        interpret_argless_method(method, args) { TypeNode.all_includers type }
```
Comment (Sija) [inline src/compiler/crystal/macros/methods.cr]: Pedantic - to match the coding style used around.

```suggestion
        interpret_argless_method(method, args) { TypeNode.all_includers(type) }
```

**Comment by RX14** [comment]
How well does this work with querying generic modules?

**Comment by Blacksmoke16** [comment]
@RX14 Seems to work fine.

```cr
module Bar(T)
end

module Foo
  include Bar(String)
end

module Baz
  include Bar(Bool)
end

{{Bar.all_includers.map &.name.stringify}} # => ["Foo", "Baz"]
```

**Comment by asterite** [comment]
I think the main issue is what happens if a module is included by a generic type. It's probably the same issue we have with `subclasses`. That's why I'm not sure how and whether this should be added to the macro methods.

**Comment by Blacksmoke16** [comment]
```cr
module Bar(T)
end

class Test(T)
  include Bar(T)
end

{{Bar.all_includers.map &.name.stringify}} # => ["Test(T)"]
```

Is what seems to happen.  Same thing as `#subclasses`.

**Comment by kalinon** [comment]
> ```crystal
> module Bar(T)
> end
> 
> class Test(T)
>   include Bar(T)
> end
> 
> {{Bar.all_includers.map &.name.stringify}} # => ["Test(T)"]
> ```
> 
> Is what seems to happen. Same thing as `#subclasses`.

I think this is acceptable behavior.

**Comment by Blacksmoke16** [comment]
Anything left to do here?

**Comment by Blacksmoke16** [comment]
Bump, this should be good to review.

File: src/compiler/crystal/macros/methods.cr [hunk 3]
```diff
@@ -1654,6 +1656,18 @@ module Crystal
       end
     end
 
+    def self.all_includers(type)
+      case type
+      when NonGenericModuleType, GenericModuleType
+        return empty_no_return_array unless types = type.raw_including_types
```
Comment (straight-shoota) [inline src/compiler/crystal/macros/methods.cr]: The main logic should not be hidden in a suffix condition. Please change this to either two separate expressions:
```suggestion
        types = type.raw_including_types
        return empty_no_return_array unless types
```
Alternatively, change the order of condition and effect:
```suggestion
        types = type.raw_including_types || return empty_no_return_array
```

File: src/compiler/crystal/macros/methods.cr [hunk 4]
```diff
@@ -1654,6 +1656,18 @@ module Crystal
...
+        return ArrayLiteral.map(types) do |including_type|
```
Comment (straight-shoota) [inline src/compiler/crystal/macros/methods.cr]: `return` is implicit here:

```suggestion
        ArrayLiteral.map(types) do |including_type|
```

**Comment by Blacksmoke16** [comment]
Also FWIW, https://github.com/crystal-lang/crystal/pull/8483 will allow you to skip the generic type vars of the module (for both this and `#subclasses/#all_subclasses` (ref: https://github.com/crystal-lang/crystal/pull/8133#issuecomment-529950943)

**Comment by Blacksmoke16** [comment]
@asterite @RX14 @bcardiff Anyone able to do another review of this?  Would be :100: if this could make it into `0.32.0`.

**Comment by RX14** [comment]
I want ary or brian to sign off on this.

**Comment by Blacksmoke16** [comment]
After working with this locally a bit I'm starting to think I should rename it to `#includers`, since this only returns types that directly include the module, not children of those types.

I'll push that tonight.

**Comment by Blacksmoke16** [comment]
@asterite @bcardiff Bump

**Pull request merged**